src/tactic/simps.lean

```diff
@@ -17,110 +17,318 @@ structures, projections, simp, simplifier, generates declarations
+        when_tracing `simps.verbose trace!"[simps] > found custom projection for {proj}:\n        > {custom_proj}",
+        return custom_proj) <|> return raw_expr,
+      mwhen (bnot <$> succeeds (is_def_eq custom_proj raw_expr))
+        fail!"Invalid custom projection:\n {custom_proj}\nExpression is not definitionally equal to {raw_expr}.",
+      return custom_proj),
+    /- check for other coercions and type-class arguments to use as projections instead. -/
+    let e_str := (expr.const str raw_levels).mk_app args,
+    raw_exprs ← automatic_projs.mfoldl (λ (raw_exprs : list expr) class_nm, (do
+      (is_class, proj_nm) ← notation_class_attr.get_param class_nm,
+      proj_nm ← proj_nm <|> (e.structure_fields_full class_nm).map list.head,
+      (raw_expr, lambda_raw_expr) ← if is_class then (do
+        guard $ args.length = 1,
+        let e_inst_type := (expr.const class_nm raw_levels).mk_app args,
+        (hyp, e_inst) ← try_for 1000 (mk_conditional_instance e_str e_inst_type),
+        raw_expr ← mk_mapp proj_nm [args.head, e_inst],
+        raw_expr_lambda ← lambdas [hyp] raw_expr, -- expr.bind_lambda doesn't give the correct type
+        return (raw_expr, raw_expr_lambda.lambdas args))
+      else (do
+        e_inst_type ← to_expr ((expr.const class_nm []).app (pexpr.of_expr e_str)),
+        e_inst ← try_for 1000 (mk_instance e_inst_type),
+        raw_expr ← mk_mapp proj_nm [e_str, e_inst],
+        return (raw_expr, raw_expr.lambdas args)),
+      raw_expr_whnf ← whnf raw_expr.binding_body,
+      let relevant_proj := raw_expr_whnf.get_app_fn.const_name,
+      guard (projs.any (= relevant_proj) ∧ ¬ e.contains (str ++ `simps ++ relevant_proj.last)),
+      let pos := projs.find_index (= relevant_proj),
+      when_tracing `simps.verbose trace!"        > using function {proj_nm} instead of the default projection {relevant_proj.last}.",
+      return $ raw_exprs.update_nth pos lambda_raw_expr) <|> return raw_exprs) raw_exprs,
+    when_tracing `simps.verbose trace!"[simps] > generated projections for {str}:\n        > {raw_exprs}",
+    simps_str_attr.set str (raw_univs, raw_exprs) tt,
+    return (raw_univs, raw_exprs)
+
+/--
+You can specify custom projections for the `@[simps]` attribute.
+To do this for the projection `my_structure.awesome_projection` by adding a declaration
+`my_structure.simps.awesome_projection` that is definitionally equal to
+`my_structure.awesome_projection` but has the projection in the desired (simp-normal) form.
+
+You can initialize the projections `@[simps]` uses with `initialize_simps_projections`
+(after declaring any custom projections). This is not necessary, it has the same effect
+if you just add `@[simps]` to a declaration.
+
+If you do anything to change the default projections, make sure to call either `@[simps]` or `initialize_simps_projections` in the same file as the structure declaration. Otherwise, you might
```

Comment:
```suggestion
If you do anything to change the default projections, make sure to call either `@[simps]` or 
`initialize_simps_projections` in the same file as the structure declaration. Otherwise, you might
```